musmart/timing/swing.py

```
@@ -0,0 +1,141 @@
+        Whether to use the log_2 of inter-onset intervals to calculate BURs, as employed in [2]. Defaults to False.
+    allow_multiple_burs_per_beat : bool, optional
+        If True, calculates a beat-upbeat ratio for every upbeat between two consecutive beats. Otherwise,
+        calculates a BUR only for the first upbeat. Defaults to False.
+
+    Returns
+    -------
+    list
+        The calculated BUR values.
+
+    Raises
+    ------
+    ValueError
+        If too few consecutive beats are provided
+
+    Examples
+    --------
+    >>> my_upbeats = np.array([0.6, 1.4])
+    >>> my_beats = np.array([0., 1., 2.])
+    >>> returned = beat_upbeat_ratio(my_beats, my_upbeats)
+    >>> [round(r, 2) for r in returned]
+    [0.67, 1.5]
+
+    >>> my_upbeats = np.array([0.6, 1.4])
+    >>> my_beats = np.array([0., 1., 2.])
+    >>> returned = beat_upbeat_ratio(my_beats, my_upbeats, use_log2_burs=True)
+    >>> [round(r, 2) for r in returned]
+    [-0.58, 0.58]
+
+    >>> my_upbeats = np.array([0.6, 0.7, 1.4])
+    >>> my_beats = np.array([0., 1., 2.])
+    >>> returned = beat_upbeat_ratio(my_beats, my_upbeats, allow_multiple_burs_per_beat=True)
+    >>> [round(r, 2) for r in returned]
+    [0.67, 0.43, 1.5]
+
+    References
+    ----------
+    [1] Benadon, F. (2006). Slicing the Beat: Jazz Eighth-Notes as Expressive Microrhythm. Ethnomusicology,
+        50/1 (pp. 73–98). https://doi.org/10.2307/20174424
+    [2] Corcoran, C., & Frieler, K. (2021). Playing It Straight: Analyzing Jazz Soloists’ Swing Eighth-Note
+        Distributions with the Weimar Jazz Database. Music Perception, 38(4), 372–385.
+        https://doi.org/10.1525/mp.2021.38.4.372
+
+    """
+
+    # Parse beats and upbeats to an array
+    beats = np.array(beats)
+    upbeats = np.array(upbeats)
+    # Remove any values from our `upbeats` variable that are also contained in `beats`
+    upbeats_ = upbeats[~np.isin(upbeats, beats)]
+    beats_ = beats[~np.isin(beats, upbeats)]
```

`isin` is rarely the right operation if you are dealing with floating point numbers. You should define a tolerance for equality

---

I see how this came from a suggestion from Mark..... but to be honest, I don't see why it's useful to have a special case for this situation. Maybe because it's ambiguous in that case which beat it corresponds to? But that depends on your matching algorithm, which I'm not clear on yet.

---

Thanks both.

My concern is to convey clearly to users what should be passed to the parameters and the relationship between them. 

Is is all event positions in one, and those assigned the role of `beats` in the other? If so, the expectation of using a strict subset suggests that it is sensible to include an `assert` check.

Depending on the algorithm (which I'm also not yet totally clear on – but intend to look through closely soon) and range of intended use cases (which is probably still open to debate), we could leave it to the user to decide whether they need/want a check on this? E.g., `check_all_in: bool = False`

Wrt tolerance for equality, perhaps try one of the `isclose` options, @HuwCheston ?

---

Yes, I think that we need clarification/discussion of the input parameters and the performed algorithm, and then this should shed light on the present question. In general though I find that silent skipping behaviour is dangerous as it often hides important problems from the user. Better to raise a warning or error, normally.

---

Presumably though if you had an 'upbeat' happening exactly on the beat that would imply that there was something wrong with your data collection

---

Should be fixed in https://github.com/music-computing/amads/pull/87. My thinking is along the lines with Peter's, that if we have an upbeat colliding exactly (or near exactly) with a beat, we should raise an error. There's a validation function `_validate_bur_inputs` that checks for these and other edge cases.